gcloud-java-bigquery/src/main/java/com/google/gcloud/spi/DefaultBigQueryRpc.java

```diff
@@ -0,0 +1,421 @@
+
+import java.io.IOException;
+import java.math.BigInteger;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class DefaultBigQueryRpc implements BigQueryRpc {
+
+  public static final String DEFAULT_PROJECTION = "full";
+  // see: https://cloud.google.com/bigquery/troubleshooting-errors
+  private static final Set<Integer> RETRYABLE_CODES = ImmutableSet.of(503);
+  private final BigQueryOptions options;
+  private final Bigquery bigquery;
+
+  public DefaultBigQueryRpc(BigQueryOptions options) {
+    HttpTransport transport = options.httpTransportFactory().create();
+    HttpRequestInitializer initializer = options.httpRequestInitializer();
+    this.options = options;
+    bigquery = new Bigquery.Builder(transport, new JacksonFactory(), initializer)
+        .setRootUrl(options.host())
+        .setApplicationName(options.applicationName())
+        .build();
+  }
+
+  private static BigQueryException translate(IOException exception) {
+    BigQueryException translated;
+    if (exception instanceof GoogleJsonResponseException) {
+      translated = translate(((GoogleJsonResponseException) exception).getDetails());
+    } else {
+      translated =
+          new BigQueryException(BigQueryException.UNKNOWN_CODE, exception.getMessage(), false);
+    }
+    translated.initCause(exception);
+    return translated;
+  }
+
+  private static BigQueryException translate(GoogleJsonError exception) {
+    boolean retryable = RETRYABLE_CODES.contains(exception.getCode())
+        || "InternalError".equals(exception.getMessage());
+    return new BigQueryException(exception.getCode(), exception.getMessage(), retryable);
+  }
+
+  @Override
+  public Dataset getDataset(String datasetId, Map<Option, ?> options) throws BigQueryException {
+    try {
+      return bigquery.datasets()
+          .get(this.options.projectId(), datasetId)
+          .setFields(FIELDS.getString(options))
+          .setQuotaUser(QUOTA_USER.getString(options))
+          .setUserIp(USER_IP.getString(options))
+          .execute();
+    } catch (IOException ex) {
+      throw translate(ex);
+    }
+  }
+
+  @Override
+  public Tuple<String, Iterable<Dataset>> listDatasets(Map<Option, ?> options)
+      throws BigQueryException {
+    try {
+      DatasetList datasetsList = bigquery.datasets()
+          .list(this.options.projectId())
+          .setQuotaUser(QUOTA_USER.getString(options))
+          .setUserIp(USER_IP.getString(options))
+          .setAll(Option.ALL_DATASETS.getBoolean(options))
+          .setMaxResults(MAX_RESULTS.getLong(options))
+          .setPageToken(PAGE_TOKEN.getString(options))
+          .execute();
+      Iterable<DatasetList.Datasets> datasets = datasetsList.getDatasets();
+      return Tuple.of(datasetsList.getNextPageToken(),
+          Iterables.transform(datasets != null ? datasets : ImmutableList.<DatasetList.Datasets>of(),
+              new Function<DatasetList.Datasets, Dataset>() {
```

Comment:
It's interesting that DatasetList.Datasets returns only partial information about the dataset.  We should probably document that if the user wants other information about the dataset not included in the dataset reference, that they should call "get."